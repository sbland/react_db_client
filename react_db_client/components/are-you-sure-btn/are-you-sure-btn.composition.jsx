import React from 'react';
import { AreYouSureBtn } from './are-you-sure-btn';

const defaultProps = {
  onConfirmed: () => alert('confirmed'),
  message: 'Yes',
  disabled: false,
  btnText: 'Click me',
  PopupPanel: ({ children, isOpen }) => isOpen ? children : '',
};

export const BasicAreYouSureBtn = () => <AreYouSureBtn {...defaultProps} />;
