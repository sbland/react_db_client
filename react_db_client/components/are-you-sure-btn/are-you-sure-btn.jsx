import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

export const AreYouSureBtn = ({
  btnText,
  onConfirmed,
  confirmMessage,
  className,
  disabled,
  notes,
  PopupPanel,
}) => {
  const [request, setRequest] = useState(false);
  const [areYouSure, setAreYouSure] = useState(false);

  useEffect(() => {
    if (areYouSure && request) {
      onConfirmed();
      setAreYouSure(false);
      setRequest(false);
    }
  }, [areYouSure, request, onConfirmed]);
  return (
    <>
      <PopupPanel isOpen={request && !areYouSure} handleClose={() => setRequest(false)}>
        <div>
          <h1>Are You Sure?</h1>
          <div>
            <button
              type="button"
              className="areYouSureCancel button-one"
              onClick={() => setRequest(false)}
            >
              Cancel
            </button>
            <button
              type="button"
              className="areYouSureAccept button-two"
              onClick={() => setAreYouSure(true)}
            >
              {confirmMessage}
            </button>
          </div>
          {notes && <div>{notes}</div>}
        </div>
      </PopupPanel>
      <button
        disabled={disabled}
        onClick={() => setRequest(true)}
        type="button"
        className={`areYouSureBtn ${className}`}
      >
        {btnText}
      </button>
    </>
  );
};

AreYouSureBtn.propTypes = {
  onConfirmed: PropTypes.func.isRequired,
  btnText: PropTypes.element.isRequired,
  confirmMessage: PropTypes.string,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  notes: PropTypes.string,
};

AreYouSureBtn.defaultProps = {
  className: 'button-one',
  disabled: false,
  notes: '',
  confirmMessage: 'Confirm',
};
