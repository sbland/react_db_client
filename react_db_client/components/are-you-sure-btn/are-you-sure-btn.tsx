import React from 'react';
import PropTypes from 'prop-types';
import {
  PopupPanelContext,
  PopupPanel,
  PopupContentWrap,
} from '@react_db_client/components.popup-panel-v2';

const popupId = 'areYouSureBtn_popupPanel';

export interface IAreYouSureProps {
  btnText: string | React.ReactNode;
  onConfirmed: () => void;
  confirmMessage?: string;
  className?: string;
  disabled?: boolean;
  notes?: string;
}

export const AreYouSureBtn = ({
  btnText,
  onConfirmed,
  confirmMessage,
  className,
  disabled,
  notes,
}: IAreYouSureProps) => {
  const { openPopup, closePopup } = React.useContext(PopupPanelContext);

  const handleFirstClick = () => {
    openPopup(popupId);
  };

  const handleCancel = () => {
    closePopup(popupId);
  };

  const handleAccept = () => {
    closePopup(popupId);
    onConfirmed();
  };

  return (
    <>
      <PopupPanel id={popupId}>
        <PopupContentWrap id={popupId} title="Are you sure?">
          <div className="areYouSurePanel_wrap">
            <h1>Are You Sure?</h1>
            <div>
              <button
                type="button"
                className="areYouSure_cancelBtn button-one"
                onClick={handleCancel}
              >
                Cancel
              </button>
              <button
                type="button"
                className="areYouSure_acceptBtn button-two"
                onClick={handleAccept}
              >
                {confirmMessage}
              </button>
            </div>
            {notes && <div>{notes}</div>}
          </div>
        </PopupContentWrap>
      </PopupPanel>
      <button
        disabled={disabled}
        onClick={() => handleFirstClick()}
        type="button"
        className={`areYouSureBtn ${className}`}
      >
        {btnText}
      </button>
    </>
  );
};

AreYouSureBtn.propTypes = {
  onConfirmed: PropTypes.func.isRequired,
  btnText: PropTypes.node.isRequired,
  confirmMessage: PropTypes.string,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  notes: PropTypes.string,
  PopupPanel: PropTypes.elementType,
};

AreYouSureBtn.defaultProps = {
  className: 'button-one',
  disabled: false,
  notes: '',
  confirmMessage: 'Confirm',
  PopupPanel: ({ children, isOpen }) => (isOpen ? children : ''),
};
