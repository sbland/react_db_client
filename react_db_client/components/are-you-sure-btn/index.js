export { AreYouSureBtn } from './are-you-sure-btn';
