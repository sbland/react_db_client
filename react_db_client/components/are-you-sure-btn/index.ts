export { AreYouSureBtn } from './are-you-sure-btn';
export type { IAreYouSureProps } from './are-you-sure-btn';
