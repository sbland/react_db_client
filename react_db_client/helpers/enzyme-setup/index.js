import { configure } from 'enzyme';
import { toMatchDiffSnapshot } from 'snapshot-diff';
import Adapter from 'enzyme-adapter-react-16';


expect.extend({ toMatchDiffSnapshot });
configure({ adapter: new Adapter() });