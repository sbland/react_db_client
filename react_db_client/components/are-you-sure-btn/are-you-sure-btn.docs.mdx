---
description: 'A AreYouSureBtn component.'
labels: ['ui', 'button']
---

import { AreYouSureBtn } from './are-you-sure-btn';

## React Btn with are you sure dialogue

A button that when clicked opens a popup that checks you are sure before calling onConfirm.

