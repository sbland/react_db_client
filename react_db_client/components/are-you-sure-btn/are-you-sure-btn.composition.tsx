import { PopupProvider } from '@react_db_client/components.popup-panel-v2';
import React from 'react';
import { AreYouSureBtn } from './are-you-sure-btn';

const defaultProps = {
  onConfirmed: () => alert('confirmed'),
  message: 'Yes',
  disabled: false,
  btnText: 'Click me',
  PopupPanel: ({ children, isOpen }) => (isOpen ? children : ''),
};

export const BasicAreYouSureBtn = () => {
  const [confirmed, setConfirmed] = React.useState(false);
  return (
    <>
      <PopupProvider>
        {' '}
        <AreYouSureBtn {...defaultProps} onConfirmed={() => setConfirmed(true)} />
      </PopupProvider>
      {confirmed && <p>User confirmed</p>}
    </>
  );
};
