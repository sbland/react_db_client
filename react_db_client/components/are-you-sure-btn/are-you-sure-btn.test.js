import '@samnbuk/react_db_client.testing.enzyme-setup';
import React from 'react';
import { shallow, mount } from 'enzyme';
import { AreYouSureBtn } from './are-you-sure-btn';

const onConfirmed = jest.fn();
const PopupPanel = jest.fn().mockImplementation(({ isOpen, children }) => isOpen ? children : '');

const defaultProps = {
  onConfirmed,
  message: 'Yes',
  disabled: false,
  className: 'demoClass',
  btnText: 'Click me',
  PopupPanel,
};

describe('AreYouSureBtn', () => {
  it('Renders', () => {
    shallow(<AreYouSureBtn {...defaultProps} />);
  });
  it('Matches Snapshot', () => {
    const component = shallow(<AreYouSureBtn {...defaultProps} />);
    const tree = component.debug();
    expect(tree).toMatchSnapshot();
  });
  describe('Unit Tests', () => {
    let component;
    beforeEach(() => {
      component = mount(<AreYouSureBtn {...defaultProps} />);
    });
    describe('are you sure process', () => {
      const clickBtn = (c) => {
        const btn = c.find('.areYouSureBtn');
        btn.simulate('click');
      };
      test('should open are you sure panel when btn button pressed', () => {
        let areYouSurePanel = component.find(PopupPanel);
        expect(areYouSurePanel.props().isOpen).toEqual(false);
        clickBtn(component);
        areYouSurePanel = component.find(PopupPanel);
        expect(areYouSurePanel.props().isOpen).toEqual(true);
      });
      test('should call confirm id when we click accept button in are you sure panel', () => {
        clickBtn(component);
        let areYouSurePanel = component.find(PopupPanel);
        const areYouSureAcceptBtn = areYouSurePanel.find('.areYouSure_acceptBtn');
        areYouSureAcceptBtn.simulate('click');
        areYouSurePanel = component.find(PopupPanel);
        expect(areYouSurePanel.props().isOpen).toEqual(false);
        expect(onConfirmed).toHaveBeenCalledWith();
      });
    });
  });
});
