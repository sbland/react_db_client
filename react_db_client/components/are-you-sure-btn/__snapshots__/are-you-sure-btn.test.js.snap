// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AreYouSureBtn Matches Snapshot 1`] = `
"<Fragment>
  <mockConstructor id=\\"areYouSureBtn_popupPanel\\" isOpen={false} handleClose={[Function: handleClose]}>
    <div className=\\"areYouSurePanel_wrap\\">
      <h1>
        Are You Sure?
      </h1>
      <div>
        <button type=\\"button\\" className=\\"areYouSure_cancelBtn button-one\\" onClick={[Function: onClick]}>
          Cancel
        </button>
        <button type=\\"button\\" className=\\"areYouSure_acceptBtn button-two\\" onClick={[Function: onClick]}>
          Confirm
        </button>
      </div>
    </div>
  </mockConstructor>
  <button disabled={false} onClick={[Function: onClick]} type=\\"button\\" className=\\"areYouSureBtn demoClass\\">
    Click me
  </button>
</Fragment>"
`;
